packages/nimble-components/src/table/tests/table.spec.ts

```diff
@@ -864,6 +870,465 @@ describe('Table', () => {
+            });
+
+            it('renders data hierarchically when id set after setData', async () => {
+                await connect();
+                element.parentIdFieldName = 'parentId';
+                await element.setData(hierarchicalData);
+                await waitForUpdatesAsync();
+                expect(pageObject.getRenderedRowCount()).toBe(0); // table in invalid state
+                element.idFieldName = 'id';
+                await waitForUpdatesAsync();
+                pageObject.clickCollapseAllButton();
+                await waitForUpdatesAsync();
+
+                expect(pageObject.getRenderedRowCount()).toBe(3); // there are 3 top level rows
+            });
+
+            it('renders data hierarchically when id and parentId set after setData', async () => {
+                await connect();
+                await element.setData(hierarchicalData);
+                await waitForUpdatesAsync();
+                expect(pageObject.getRenderedRowCount()).toBe(7); // there are 7 total rows
+                element.idFieldName = 'id';
+                element.parentIdFieldName = 'parentId';
+                await waitForUpdatesAsync();
+                pageObject.clickCollapseAllButton();
+                await waitForUpdatesAsync();
+
+                expect(pageObject.getRenderedRowCount()).toBe(3); // there are 3 top level rows
+            });
+
+            it('expanding parent results in rendered child rows', async () => {
+                await connect();
+                element.idFieldName = 'id';
+                element.parentIdFieldName = 'parentId';
+                await element.setData(hierarchicalData);
+                await waitForUpdatesAsync();
+                pageObject.clickCollapseAllButton();
+                await waitForUpdatesAsync();
+
+                pageObject.clickDataRowExpandCollapseButton(0); // parent has 2 children
+                await waitForUpdatesAsync();
+
+                expect(pageObject.getRenderedRowCount()).toBe(5); // 3 rows initially, two more after expansion
+            });
+
+            it('collapsing parent hides children', async () => {
+                await connect();
+                element.idFieldName = 'id';
+                element.parentIdFieldName = 'parentId';
+                await element.setData(hierarchicalData);
+                await waitForUpdatesAsync();
+                pageObject.clickCollapseAllButton();
+                await waitForUpdatesAsync();
+
+                pageObject.clickDataRowExpandCollapseButton(0); // expands parent
+                await waitForUpdatesAsync();
+                expect(pageObject.getRenderedRowCount()).toBe(5);
+
+                pageObject.clickDataRowExpandCollapseButton(0); // collapses parent
+                await waitForUpdatesAsync();
+
+                expect(pageObject.getRenderedRowCount()).toBe(3);
+            });
+
+            it('maintains expanded state between data updates', async () => {
```

Comment:
What is the behavior if a record is in the data and expanded, then a data update deletes it, then another data update restores it? Is its original expanded state maintained? If "yes" then we should probably have a test for it. If "no" then we probably don't need one for that case, but we should make sure that we have a test asserting that new data is always rendered in the same default expanded state (presumably collapsed).

Reply:
So, in this particular test new data is being added to the existing set of data, and we ultimately assert two separate things:
1. the new data added to the existing data defaults to expanded (which is the expected default expanded state...not collapsed)
2. that the previous row that was expanded (after first collapsing everything), stays expanded after adding the new records.

To answer your other question, yes, if a record is removed and then re-added it will be rendered with the default expanded state (which is expanded). I think there actually is value in writing a test for that.